runtime/compiler/ilgen/Walker.cpp

```diff
@@ -6079,6 +6079,22 @@ TR_J9ByteCodeIlGenerator::loadFromCallSiteTable(int32_t callSiteIndex)
 void
 TR_J9ByteCodeIlGenerator::loadArrayElement(TR::DataType dataType, TR::ILOpCodes nodeop, bool checks)
    {
+   if (TR::Compiler->om.areValueTypesEnabled() && dataType == TR::Address)
```

Comment:
How come we don't check if the array element is value type first and then if it's flattened before going down this path to call VM helper to load flattened array element?

Reply:
`aaload` and `aastore` don't use the constant pool, so the type of array elements is not known (at least, not without some dataflow analysis). The only thing that's known is that it's an array of some object type.